src/java/org/apache/cassandra/db/filter/IndexHints.java

```diff
@@ -0,0 +1,674 @@
+    }
+
+    /**
+     * @param indexName the name of an index
+     * @return {@code true} if the index is excluded, {@code false} otherwise
+     */
+    public boolean excludes(String indexName)
+    {
+        for (IndexMetadata i : excluded)
+        {
+            if (i.name.equals(indexName))
+                return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param indexes a set of indexes
+     * @return the indexes that are not excluded by these hints
+     */
+    public <T extends Index> Set<T> notExcluded(Iterable<T> indexes)
+    {
+        Set<T> result = new HashSet<>();
+        for (T index : indexes)
+        {
+            if (!excludes(index))
+                result.add(index);
+        }
+        return result;
+    }
+
+    /**
+     * Returns the best of the specified indexes that satisfies the specified filter, according to these hints first,
+     * and the index-provided selectivity second.
+     *
+     * @param indexes a collection of indexes
+     * @param filter a filter to apply to the indexes
+     * @param isContains whether the calling operation is a {@code [NOT] CONTAINS [KEY]} operation, in which case we
+     * prefer not-analyzed indexes (see CNDB-13925).
+     * @return the best of the specified indexes that satisfies these index hints and the specified filter
+     */
+    public <T extends Index> Optional<T> getBestIndexFor(Collection<T> indexes, Predicate<T> filter, boolean isContains)
+    {
+        // filter excluded and filtered indexes
+        Collection<T> candidates = filter(indexes, index -> !excludes(index) && filter.test(index));
+
+        // prefer included indexes
+        candidates = prefer(candidates, this::includes);
+
+        // if we are using a contains operator, we prefer indexes without an analyzer (see CNDB-13925)
+        if (isContains)
+            candidates = prefer(candidates, index -> !index.isAnalyzed());
+
+        // return the candidate with the best selectivity
+        return bestSelectivityIndex(candidates);
+    }
+
+    /**
+     * Returns the indexes in the specified collection of indexes that are applicable to the specified filter.
+     *
+     * @param indexes a collection of indexes
+     * @param filter a filter to apply to the indexes
+     * @return the indexes that are applicable to the specified filter
+     */
+    private static <T extends Index> Collection<T> filter(Collection<T> indexes, Predicate<T> filter)
+    {
+        if (indexes.isEmpty())
+            return indexes;
+
+        Set<T> candidates = new HashSet<>(indexes.size());
+        for (T index : indexes)
+        {
+            if (filter.test(index))
+                candidates.add(index);
+        }
+        return candidates;
+    }
+
+    /**
+     * Returns the indexes in the specified collection that are preferred by the specified predicate, or the unmodified
+     * collection if there are no preferred indexes.
```

Comment:
I am not sure if prefer predicate can be confusing, thus
```suggestion
     * Returns the indexes in the specified collection that satisfy the specified predicate, or the unmodified
     * collection if there are no indexes satisfying the predicate.
```